packages/jsii/lib/assembler.ts

```diff
@@ -311,13 +327,105 @@ export class Assembler implements Emitter {
+      // Nothing to do here...
+      return;
+    }
+    const moduleSpecifier = declaration.parent.moduleSpecifier;
+    if (moduleSpecifier == null || !ts.isStringLiteral(moduleSpecifier)) {
+      // There is a grammar error here, so we'll let tsc report this for us.
+      return;
+    }
+    const resolution = ts.resolveModuleName(
+      moduleSpecifier.text,
+      declaration.getSourceFile().fileName,
+      this.program.getCompilerOptions(),
+      ts.sys
+    );
+    if (resolution.resolvedModule == null) {
+      // Unresolvable module... We'll let tsc report this for us.
+      return;
+    }
+    if (resolution.resolvedModule.isExternalLibraryImport) {
+      // External re-exports are "pure-javascript" sugar; they need not be
+      // represented in the jsii Assembly since the types in there will be
+      // resolved through dependencies.
+      return;
+    }
+    const sourceFile = this.program.getSourceFile(resolution.resolvedModule.resolvedFileName)!;
+    const sourceModule = this._typeChecker.getSymbolAtLocation(sourceFile);
+    // If there's no module, it's a syntax error, and tsc will have reported it for us.
+    if (sourceModule) {
+      this._addToNamespace(symbol, sourceModule);
+    }
+  }
+
+  private _addToNamespace(ns: ts.Symbol, moduleLike: ts.Symbol) {
+    for (const symbol of this._typeChecker.getExportsOfModule(moduleLike)) {
+      if (this._namespaceMap.has(symbol)) {
+        const currNs = this._namespaceMap.get(symbol)!;
+        if (currNs.name !== ns.name) {
+          const currNsDecl = currNs.valueDeclaration ?? currNs.declarations[0];
+          const nsDecl = ns.valueDeclaration ?? ns.declarations[0];
+          this._diagnostic(
+            symbol.valueDeclaration,
+            ts.DiagnosticCategory.Error,
+            `Symbol is re-exported under two distinct namespaces (${currNs.name} and ${ns.name})`,
+            [{
+              category: ts.DiagnosticCategory.Warning,
+              file: currNsDecl.getSourceFile(),
+              length: currNsDecl.getStart() - currNsDecl.getEnd(),
+              messageText: `Symbol is exported under the "${currNs.name}" namespace`,
+              start: currNsDecl.getStart(),
+              code: JSII_DIAGNOSTICS_CODE
+            }, {
+              category: ts.DiagnosticCategory.Warning,
+              file: nsDecl.getSourceFile(),
+              length: nsDecl.getStart() - nsDecl.getEnd(),
+              messageText: `Symbol is exported under the "${ns.name}" namespace`,
+              start: nsDecl.getStart(),
+              code: JSII_DIAGNOSTICS_CODE
+            }]
+          );
+        }
+        // Found two re-exports, which is odd, but they use the same namespace,
+        // so it's probably okay? That's likely a tsc error, which will have
+        // been reported for us already anyway.
+        continue;
+      }
+      this._namespaceMap.set(symbol, ns);
+
+      const decl = symbol.declarations?.[0];
+      if (decl != null
```

Comment:
add a comment such as "if this is a supported jsii type"...

Also, do we want a warning if it's not?

Reply:
I don't think the warning is useful (exports that cannot be modeled by *jsii* are logged somewhere else as part of the next pass).